### PR TITLE
Fix SQLite3 quoting to not put database and table in separate quotes

### LINKED ISSUE #35292: SQLite no longer works when using table_name with a dot ex: "foo.bar"
### Steps to reproduce
The following fails on master.
[sqlite_master_demo.txt](https://github.com/rails/rails/files/2871394/sqlite_master_demo.txt)
But the same code ran against rails 5.2, and all version since 4, will pass:
[sqlite_521_demo.txt](https://github.com/rails/rails/files/2871396/sqlite_521_demo.-.Copy.txt)

Without using a complicated file (or creating any table), the following makes it easy to try:
```
class Post < ActiveRecord::Base
  self.table_name = "bar.posts"
end
Post.first
```
In 6.0.0, SQLite whines of a syntax error, while on 5.2.1, it whine of a table that doesn't exist (as expected).

### Expected behavior
When using a "schema" or an "attached database", (names change based on which DB), things normally just work by using a simple `self.table_name = "bar.posts"`. This is no longer the case for SQLite3 in rails 6.0.0.

For the example with Post.first above, the generated SQL should be:
`SELECT  "foo.bar".* FROM "foo.bar" ORDER BY "foo.bar"."id" ASC LIMIT ?`

### Actual behavior
Since 6.0.0, it now does:
`SELECT "foo"."bar".* FROM "foo"."bar" ORDER BY "foo"."bar"."id" ASC LIMIT ?`

### System configuration
**Rails version**: 6.0.0.beta1

**Ruby version**: 2.5.0

### More details
This is the PR that introduced this bug: https://github.com/rails/rails/pull/34626

### PR DESCRIPTION
**Pull request by MaxLap**
This reverts a fix that broke schemas in SQLite3, and applies a different fix to solve the same test case.

This fixes #35292.

Basically, when passing a symbol with a dot to `pluck`, we are expecting it to be a table + a column. MySQL and PostgreSQL handled it fine, because quoting `schema + table` is the same as quoting `table + column`: `"part1"."part2"`. So a quote_table_name was being applied by pluck.

However, this is not the case in SQLite3, which wants `"schema.table"."column"` (note the schema and table in the same quotes), and so would generate `"table.column"`. The applied logic is instead to check for a dot, and if found, to quote each part separately.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/sqlite3/quoting.rb [hunk 1]
```diff
@@ -12,10 +12,6 @@ def quote_table_name_for_assignment(table, attr)
           quote_column_name(attr)
         end
 
-        def quote_table_name(name)
```
Comment (gmcgibbon) [inline activerecord/lib/active_record/connection_adapters/sqlite3/quoting.rb]: This removes memoization, which I think we don't want. Please revert this line and only remove the `gsub`.

File: activerecord/test/cases/adapters/sqlite3/schema_test.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+class SQLite3SchemaTest < ActiveRecord::SQLite3TestCase
+
+  class SchemaThing < ActiveRecord::Base
+    self.table_name = "test_schema.things"
+  end
+
+  def setup
+    original_verbose = ActiveRecord::Migration.verbose
+    ActiveRecord::Migration.verbose = false
+    ActiveRecord::Schema.define do
```
Comment (gmcgibbon) [inline activerecord/test/cases/adapters/sqlite3/schema_test.rb]: I would say this goes against the current test suite's standards. You should be able to run tests on any DB by changing `ARCONN` and using `if` statements to not run tests that don't apply to specific adapters. Attaching a DB also doesn't seem to work in CI. Can you refactor this to use the standard test schema?

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 3]
```diff
@@ -1055,7 +1055,13 @@ def arel_columns(columns)
           case field
           when Symbol
             field = field.to_s
-            arel_column(field) { connection.quote_table_name(field) }
+            arel_column(field) do
+              table, sep, column = field.to_s.rpartition(".")
```
Comment (gmcgibbon) [inline activerecord/lib/active_record/relation/query_methods.rb]: This only accounts for a single dot, but you can have multiple dots in a table name, eg. `field = :"a.b.c"`

**Review by gmcgibbon — CHANGES_REQUESTED** [review]

File: activerecord/lib/active_record/connection_adapters/sqlite3/quoting.rb (hunk 1 above)
Comment (MaxLap) [reply activerecord/lib/active_record/connection_adapters/sqlite3/quoting.rb]: I undo things to exactly how things were before your fix. `supers just ends up calling `quote_column_name`, which does cache things. I don't mind changing it back if you prefer. Let me know.

File: activerecord/lib/active_record/relation/query_methods.rb (hunk 3 above)
Comment (MaxLap) [reply activerecord/lib/active_record/relation/query_methods.rb]: This uses `rpartition`, so it splits from the last dot.
For your example, the result would be `['a.b', '.', 'c']`

File: activerecord/test/cases/adapters/sqlite3/schema_test.rb (hunk 2 above)
Comment (MaxLap) [reply activerecord/test/cases/adapters/sqlite3/schema_test.rb]: Weird, the whole test suite passed on my rails-dev-box. I will not have time to look into it until tomorrow.

There are plenty of tests that are specific to one adapter like that. I basically copied the `schema_test.rb` file name and basic structure from the postgresql side of things: https://github.com/rails/rails/blob/master/activerecord/test/cases/adapters/postgresql/schema_test.rb, there is also an equivalent file in the mysql side.

I'm not quite sure what is wrong here to be honest. I work a lot by following examples I find in the project that match my need. Could you show me what you expect?

I agree a single set of tests to checks schema's behaviors, with conditions to setup the schema correctly for each databse would be optimal, thats what I do in the gem that found this issue. But I'm not confortable doing a refactor of hundreds of lines (all 3 databases's schema_test files) of test code in this massive test suite.

File: activerecord/lib/active_record/relation/query_methods.rb (hunk 3 above)
Comment (kamipo) [reply activerecord/lib/active_record/relation/query_methods.rb]: Can you fix it in the `quote_table_name`?
Here is not only place we use `quote_table_name`.

Comment (kamipo) [reply activerecord/lib/active_record/relation/query_methods.rb]: hm.. I see the problem.

https://github.com/rails/rails/blob/825d70b4069df3411b77636c98ca668292bbc03a/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L42-L44

We should use `quote_table_name` only to quote actual table name or schema qualified table name.
Here is one of the case of quoting table name qualified *column name*.
We probably need new method for the purpose.

Comment (MaxLap) [reply activerecord/lib/active_record/relation/query_methods.rb]: I can transfer my logic to `Quoting` under the name `quote_table_and_column_name`.

The issue is how to deal with column names which might also contain dots. My fix doesn't deal with those, that's why I didn't want to make a method; it wasn't as general purpose as I would like.

File: activerecord/lib/active_record/connection_adapters/sqlite3/quoting.rb (hunk 1 above)
Comment (gmcgibbon) [reply activerecord/lib/active_record/connection_adapters/sqlite3/quoting.rb]: Yes, [abstract quoting](https://github.com/rails/rails/blob/dfa857ef25b5abf7485e4f7115e4a311e00c599c/activerecord/lib/active_record/connection_adapters/abstract/quoting.rb#L72) calls `quote_column_name ` but it uses the wrong ivar. I would prefer keeping them separate as they were before.

File: activerecord/lib/active_record/relation/query_methods.rb (hunk 3 above)
Comment (kamipo) [reply activerecord/lib/active_record/relation/query_methods.rb]: At least the `quote_table_name` is also used for quoting the order by column.
4650874073bef4f02da3c2b0bfc4ba2a6fd5d562 would be broken by this change for now.

File: activerecord/lib/active_record/connection_adapters/sqlite3/quoting.rb (hunk 1 above)
Comment (MaxLap) [reply activerecord/lib/active_record/connection_adapters/sqlite3/quoting.rb]: Okay, but then do I use `super` or not? Because if I do, then both ivar will cache the value which feels wasteful.

File: activerecord/test/cases/adapters/sqlite3/schema_test.rb (hunk 2 above)
Comment (MaxLap) [reply activerecord/test/cases/adapters/sqlite3/schema_test.rb]: PostgreSQL does create an actual schema for its tests, so in my mind I should do the same. 

I understand that you don't want me to use `ActiveRecord::Schema`. I can look at doing it from the connection object as you ask. That was the first thing I did, but I had the weird issue where `Thing.first` and `Thing.create` would return a `Thing` that had nil as id. This gets solved by using the Schema. I guess it's another bug down in the rabbit hole.

I'll look into it tomorrow.

File: activerecord/lib/active_record/relation/query_methods.rb (hunk 3 above)
Comment (MaxLap) [reply activerecord/lib/active_record/relation/query_methods.rb]: Alright, thnx, I'll rebase and fix+test again when I move the logic to Quoting.

**Comment by MaxLap** [comment]
Okay, so the root cause of all of this is absolutely bonkers. Nothing is what it seems and I was wrong. But the story is interesting.

TL;DR: Quoting schema/tables is the same in SQLite as the other databases. No weird need for ".", the fix is correct and just made my code start actually testing the right thing and find a subtle bug elsewhere. SQLite's Attached database was never actually supported, and supporting them appears not worth the trouble.

Because of the bad quoting that was going on before the fix, when my gem, in its tests, would create a table in a database (using `create_table "my_schema.thing" do ...`), the actual generated SQL would be a `CREATE TABLE "my_schema.thing"`). This is actually a table with a name that contain a dot, and not a table that refers to the schema.

Then, later on, the code in all my tests also also used the quote_table_name, and so everything always used a table_name with a dot instead of testing actual schemas.

What threw me in this weird quoting behavior understanding in 6.0 is that I would get syntax errors when trying to use the models. But that was not form the select or such things. No. It was from a call to PRAGMA which does things special for schemas (attached database). See, this is the query to get the table information in SQLite:

`"PRAGMA table_info(#{quote_table_name(table_name)})"`

However, when dealing with database, the actual syntax is

`"PRAGMA schema_name.table_info(table_name)"`

But those kinds of calls are not shown in the log, so I would get confused thinking the problem was in the previous query (because the problematic query is not shown in the error messages... which i don't know why).

Now that part could be fixed with some code to split the table name into schema + table in an intelligent way (so that quoted parts remain the same). But here is a the kicker:

In a `SELECT` in SQLite, you cannot specify the database at the attributes parts. Meaning:

`SELECT bar.id FROM foo.bar` is valid, but
`SELECT foo.bar.id FROM foo.bar` is not valid, its a syntax error!

So in order to generate correct queries here, things get even more wild. It's clear to me that this is not worth my trouble (or anyone's).

Sorry for the waste of time. Such a rabbit hole...
